### DISCUSSION TITLE
Custom Controls and Data Bindings

### DISCUSSION THREAD

**Discussion by ajsuydam** [dotnet/maui#23280] [Q&A]
Hey y'all. I know this is probably the thousandth time someone has asked a question like this, but I have done a lot of digging and cannot seem to find a solution to what is apparently a problem that is unique to me (please prove me wrong, this is driving me nuts).

I have a custom control (XAML w/ C# code behind). There is a bindable property on that control. The UI for the control references values stored in that property for display through data binding.

That bindable property can be bound to by the page that consumes the control, such that the view model for the page can be used as the source for the object that goes into that binding.

However, when I update the property on the view model with some new data, the UI doesn't reflect the change. I have tried every binding mode and source and reference known to man, and cannot seem to get it working. If this isn't enough detail, please let me know and I'll provide a MRE (minimum repro example).

**Comment by OvrBtn** [comment]
Hi there, I think it will be simpler if you provide some small reproduction repo or at least some code. It might be something small but it's hard to guess like that 😉

**Comment by ajsuydam** [comment]
@OvrBtn Yeah, I can definitely try!

So this is something like my MainPage.xaml, I have a view model for the page, and a custom control, a property of which is bound to some property on the view model.
```
 - header w/namespaces -
 xmlns:local=MyApp.CustomControls
x:DataType=MyPageVM

    <local:MyControl LabelText="{Binding FromViewModel}"/>


```

This is the view model (MyPageVM.cs) with that property (I'm using the communitytoolkit.mvvm helpers here, not sure if that is part of the issue):
```
[ObservableProperty]
public string FromViewModel { get; set; };

```

Then I have my custom control defined something like this:

MyControl.xaml

```
<ContentView ...>

    <Label Text="{Binding LabelText}"/>


</ContentView>


```
MyControl.xaml.cs:

```
public static readonly BindableProperty LabelTextProperty = new BindableProperty.Create(
    nameof(LabelText),
    typeof(string),
.......

public string labelText { get; set; } = "";

//Ctor, other methods

```

**Comment by OvrBtn** [comment]
I took a quick look and I'm seeing 2 issues:
1. you are using [ObservableProperty] wrong refer to [docs](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/observableproperty) for details but basically how you should do this is:
instead of your 
```
[ObservableProperty]
public string FromViewModel { get; set; };
```
it shoudl be
```
[ObservableProperty]
string fromViewModel;
```
The public property being "public string FromViewModel {get;set;}" is automatically generated by CommunityToolkit with all the logic to invoke OnPropertyChanged when needed, that's the whole point of using this attribute. IntelliSense should tell you that what you've made is wrong. Also remember that when using [ObservableProperty] you class has to be partial.

2. You are creating BindableProperty wrong instead of
```
public static readonly BindableProperty LabelTextProperty = new BindableProperty.Create(
    nameof(LabelText),
    typeof(string),
.......

public string labelText { get; set; } = "";
```
it should be
```
public static readonly BindableProperty LabelTextProperty = BindableProperty.Create(
    nameof(LabelText),
    typeof(string),
.......

public string LabelText
{
   get => (string)GetValue(LabelTextProperty);
   set => SetValue(LabelTextProperty, value);
}
```
Checkout [docs](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/bindable-properties?view=net-maui-8.0#create-accessors)

**Comment by ajsuydam** [comment]
@OvrBtn thanks for the detailed reply. I was rushing some stuff on friday, so I'm sorry the example wasn't totally complete. The edits you suggested I actually already had in my code (which I should have referenced rather than trying to do from memory). The bindable properties have backing fields, and the ObservableProperty fields are private and lowerCamelCase already.

I also should add that the type of the property is a custom object that I have defined, so it is a non-trivial type. I don't know if that would make a difference either.

**Comment by OvrBtn** [comment]
> @OvrBtn thanks for the detailed reply. I was rushing some stuff on friday, so I'm sorry the example wasn't totally complete. The edits you suggested I actually already had in my code (which I should have referenced rather than trying to do from memory). The bindable properties have backing fields, and the ObservableProperty fields are private and lowerCamelCase already.
> 
> I also should add that the type of the property is a custom object that I have defined, so it is a non-trivial type. I don't know if that would make a difference either.

That's okay.

In terms of the non trivial type, it really depends on what you do. For example let's imagine that FromViewModel is not type of string but some more complex class so if you do `FromViewModel = new SomeKindOfClass()` then CommunityToolkit should trigger OnPropertyChanged and UI binded to this property should change but if you do `FromViewModel.SomeProperty = someNewValue` OnPropertyChanged probably won't be automatically (automatically meaning by CommunityToolkit) triggered.

**Reply by ajsuydam** [reply]
OH.

Yes this is it. I'm trying to update some sub-field of the object. So this is a CommunityToolkit limitation you think?

**Reply by OvrBtn** [reply]
> OH.
> 
> Yes this is it. I'm trying to update some sub-field of the object. So this is a CommunityToolkit limitation you think?

Honestly I don't know the exact implementation of [ObservableProperty] to judge that, try implementing `INotifyPropertyChanged` in your class and then try invoking PropertyChanged when property is set, check out example in [docs](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/binding-mode?view=net-maui-8.0#viewmodels-and-property-change-notifications)

**Comment by ajsuydam** [comment]
GAH. This is so frustrating. I've somehow tied myself into a knot, I'm sure what I am doing wrong is super simple.

Crucial detail I'm finding: The property is updated in my Model, not the view model. I believe my view model actually holds a reference to the data in the model. When that data is changed (in the model) the UI doesn't reflect that change. That is where I get stuck. In the model we are updating some sub-field of the `SomeKindOfClass` instance (which is itself contained in a owning object). References and C# are weird man. If you have any thoughts on this, lemme know. I may just need to rip all of this out and start over.
